Feature: Custom Entity API

onebeastchris opened the pull request:
Introducing: Custom Entities!

*tl;dr: with this PR, you can summon custom bedrock entities that replace the vanilla mapping*

# Additions 
- `CustomEntityDefinition` / `GeyserEntityDefinition`: Representations of custom, and vanilla Bedrock entities. Unlike custom blocks/items, bedrock entities have fewer properties that are defined in advance; it's just the identifier, and the Bedrock Entity Properties. Setting entity properties was introduced in a previous PR; and works the same way.

- `JavaEntityType`: Represents a vanilla Java entity type, with the width / height / type identifier, as well as the default Bedrock entity associated with it. ~~Similarly, `CustomJavaEntityType` represents a non-vanilla Java entity - however, that part of the API still needs some more work and should be regarded as unstable.~~ postponed indefinitely

- `GeyserEntityDataType` / `GeyserEntityDataTypes`: These are representations of various Bedrock entity metadata types, such as scale, width, height, variant, or color. Further, `vertical_offset` has been added as a "custom" data type to allow setting a vertical entity offset.

- The `GeyserEntity` class has seen major additions! You can now query the entities' associated Bedrock entity definition, Java position, the Geyser id, UUID, or update / query the aforementioned data types.

- You can now look up `GeyserEntity` instances using the entity UUID or Geyser entity ID, additionally to the Java entity id.

# New Events
- `GeyserDefineEntitiesEvent`: Allows registering custom Bedrock entities and querying existing entities.

- `SessionSpawnEntityEvent`: Base entity spawn event extended by the server events. With it, you can set a pre-spawn consumer, and switch the Bedrock entity definition, or cancel the entity spawn outright.

- `ServerAttachParrotsEvent`: Called every time a parrot is spawned on the player entity

- `ServerSpawnEntityEvent`: Called for every non-player entity that is spawned by the Java server. Within this event, you can query the Java entity type, uuid, and entity id (and also have access to the methods provided by the `SessionSpawnEntityEvent`!

Here's some example code of the API in action:
https://gist.github.com/onebeastchris/1521ab585669792a79a9558d9d069834

<img width="516" height="309" alt="image" src="https://github.com/user-attachments/assets/433c70a0-b00c-48ff-92ac-ddfa72ce2522" />
<img width="148" height="184" alt="image" src="https://github.com/user-attachments/assets/2daf0210-d949-49b4-9380-0e1b73439865" />

Internal changes:
- Bedrock entity definitions are now split from the Java entity definitions. 
- There is now a `EntitySpawnContext` that is passed in entity constructors - instead of many arguments. This should make it easier to add new entities, call events, or add more arguments in the future
- Entity offsets are now properly handled in the base `Entity` class - no more hacks in TntEntity and the like!

TO-DO's:
- [x] Debug logging for modifications
- [x] Docs

*EXPERIMENTAL* downloads:
-  [Geyser-Bungeecord.jar](https://download.geysermc.org/v2/projects/geyserpreview/versions/pr.6010/builds/latest/downloads/bungeecord)
- [Geyser-Fabric.jar](https://download.geysermc.org/v2/projects/geyserpreview/versions/pr.6010/builds/latest/downloads/fabric)
- [Geyser-NeoForge.jar](https://download.geysermc.org/v2/projects/geyserpreview/versions/pr.6010/builds/latest/downloads/neoforge)
- [Geyser-Spigot.jar](https://download.geysermc.org/v2/projects/geyserpreview/versions/pr.6010/builds/latest/downloads/spigot)
- [Geyser-Standalone.jar](https://download.geysermc.org/v2/projects/geyserpreview/versions/pr.6010/builds/latest/downloads/standalone)
- [Geyser-Velocity.jar](https://download.geysermc.org/v2/projects/geyserpreview/versions/pr.6010/builds/latest/downloads/velocity)
- [Geyser-ViaProxy.jar](https://download.geysermc.org/v2/projects/geyserpreview/versions/pr.6010/builds/latest/downloads/viaproxy)

eclipseisoffline committed:
Start working on custom entity API (for real this time)

eclipseisoffline committed:
Split EntityDefinition up into EntityDefinition and VanillaEntityDefinition, create GeyserEntityType, target custom entities MCPL

eclipseisoffline committed:
Allow creating custom entity types from API

eclipseisoffline committed:
Further refactor EntityDefinition, create a Base class

eclipseisoffline committed:
Work on exposing custom entity definitions in the API

eclipseisoffline committed:
Add EntityDefinition#is

eclipseisoffline committed:
Implement CustomEntityDefinition, part one

onebeastchris committed:
Move CustomEntityDefinition into custom package

onebeastchris committed:
CustomEntityDefinition should extend GeyserEntityDefinition

onebeastchris committed:
Separate BedrockEntityDefinition from Entity classes

onebeastchris committed:
EntityDefinition -> EntityTypeDefinition

onebeastchris committed:
Continue splitting bedrock entities from java types

onebeastchris committed:
Width, height and offset should be tied to the Java entity type; and can now be modified in the ServerSpawnEntityEvent

onebeastchris committed:
More work on events, closer to working

onebeastchris committed:
Fix startup

onebeastchris committed:
Expose geyserId in API, add SessionAttachParrotsEvent

onebeastchris committed:
Rename SessionAttachParrotsEvent -> ServerAttachParrotsEvent, fill in API Javadocs

onebeastchris commented on core/src/main/java/org/geysermc/geyser/translator/protocol/java/entity/JavaSetPassengersTranslator.java:
```diff
@@ -59,7 +59,7 @@ public void translate(GeyserSession session, ClientboundSetPassengersPacket pack
                 session.getPlayerEntity().setVehicle(entity);
                 // We need to confirm teleports before entering a vehicle, or else we will likely exit right out
-                session.confirmTeleport(passenger.getPosition().down(EntityDefinitions.PLAYER.offset()));
+                session.confirmTeleport(passenger.getPosition().down(VanillaEntities.PLAYER_ENTITY_OFFSET));
```
no, it is session player entity here, never null

onebeastchris committed:
Apply some suggestions

onebeastchris committed:
Another couple changes, start designing an entity data system to set scale/variant/height/width and so forth

Camotoy commented on api/src/main/java/org/geysermc/geyser/api/entity/EntityData.java:
```diff
@@ -49,6 +49,16 @@ public interface EntityData {
+    /**
+     * Returns a {@link GeyserEntity} to e.g. update entity properties.
+     */
+    @NonNull CompletableFuture<@Nullable GeyserEntity> entityByUuid(@NonNegative UUID javaUuid);
```
These don't need to be CompletableFutures because get checks on Fastutil maps can be concurrent. Modifying them cannot be done off-thread. Not sure how to deal with an API break like the one above - it was absolutely a terrible decision to make early on because it's simply unneeded.

onebeastchris committed:
Api design goes brrr

onebeastchris committed:
Implement dirty entities to send updates once a tick

onebeastchris committed:
Fix NPE in test

onebeastchris commented on api/src/main/java/org/geysermc/geyser/api/entity/custom/CustomJavaEntityType.java:
```diff
@@ -0,0 +1,89 @@
+         * @param defaultBedrockDefinition the default Bedrock definition
+         * @return this builder
+         */
+        Builder defaultBedrockDefinition(@Nullable GeyserEntityDefinition defaultBedrockDefinition);
```
todo - just bedrockDefinition? Or geyserDefinition? Also used in other places

onebeastchris committed:
Attempt to add feature branch publishing

onebeastchris committed:
API polishing; properly initialize height/width

onebeastchris committed:
Refactor: unify offset handling, always store Java entity position

onebeastchris committed:
Remove unused offsets

onebeastchris committed:
Remove debug print, validate that entity definitions have been registered, setup publishing

onebeastchris committed:
Set up feature branch publishing and vertical offset applying

onebeastchris committed:
correct javadoc

onebeastchris committed:
Merge branch 'master' into feature/custom-entities-api

onebeastchris committed:
Merge branch 'master' into feature/custom-entities-api

onebeastchris committed:
Merge remote-tracking branch 'upstream/feature/custom-entities-api' into this-is-the-entity-api

onebeastchris committed:
Merge remote-tracking branch 'upstream/master' into this-is-the-entity-api

# Conflicts:
#	core/src/main/java/org/geysermc/geyser/entity/type/player/MannequinEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/player/SkullPlayerEntity.java

onebeastchris committed:
Merge branch 'master' into feature/custom-entities-api

QuickGlare committed:
Change entity definition log level to debug for missing definitions (#6042)

onebeastchris committed:
Fix villager inventories not showing up

onebeastchris committed:
Fix interaction range check

onebeastchris committed:
Merge remote-tracking branch 'upstream/master' into feature/custom-entities-api

# Conflicts:
#	core/src/main/java/org/geysermc/geyser/entity/type/BoatEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/Entity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/living/animal/horse/AbstractHorseEntity.java
#	core/src/main/java/org/geysermc/geyser/translator/protocol/bedrock/entity/player/input/BedrockPlayerAuthInputTranslator.java

onebeastchris committed:
Expose entity hitboxes in API, store bedrock position to avoid re-calculation

onebeastchris committed:
Minor cleanup

onebeastchris committed:
Log unknown implementations of GeyserEntityDefinition

onebeastchris committed:
Use position setter instead of directly setting entity position

onebeastchris committed:
Fix client crashing related to sending & unlocking smithing recipes

onebeastchris committed:
Merge in upstream, we are on 1.21.11 now

onebeastchris committed:
Merge remote-tracking branch 'upstream/master' into feature/custom-entities-api

# Conflicts:
#	core/src/main/java/org/geysermc/geyser/entity/type/Entity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/LivingEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/player/AvatarEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/player/MannequinEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/player/SkullPlayerEntity.java
#	core/src/main/java/org/geysermc/geyser/skin/SkinManager.java
#	core/src/main/java/org/geysermc/geyser/translator/protocol/bedrock/BedrockNetworkStackLatencyTranslator.java
#	core/src/main/java/org/geysermc/geyser/translator/protocol/java/entity/JavaRemoveMobEffectTranslator.java
#	core/src/main/java/org/geysermc/geyser/util/DimensionUtils.java

onebeastchris committed:
Merge upstream

onebeastchris committed:
Merge remote-tracking branch 'upstream/master' into feature/custom-entities-api

# Conflicts:
#	core/src/main/java/org/geysermc/geyser/entity/VanillaEntities.java
#	core/src/main/java/org/geysermc/geyser/entity/spawn/EntitySpawnContext.java
#	core/src/main/java/org/geysermc/geyser/entity/type/AbstractArrowEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/BoatEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/Entity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/FireballEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/FishingHookEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/InteractionEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/LeashKnotEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/LivingEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/MinecartEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/TNTEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/TextDisplayEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/ThrowableEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/ThrowableItemEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/WitherSkullEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/living/AbstractFishEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/living/AgeableEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/living/ArmorStandEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/living/MagmaCubeEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/living/SquidEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/living/animal/SnifferEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/living/animal/horse/CamelEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/living/merchant/AbstractMerchantEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/living/monster/EnderDragonEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/living/monster/EnderDragonPartEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/living/monster/ShulkerEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/living/monster/raid/SpellcasterIllagerEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/player/AvatarEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/player/PlayerEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/player/SessionPlayerEntity.java
#	core/src/main/java/org/geysermc/geyser/level/physics/CollisionManager.java
#	core/src/main/java/org/geysermc/geyser/session/cache/EntityCache.java
#	core/src/main/java/org/geysermc/geyser/session/cache/FormCache.java
#	core/src/main/java/org/geysermc/geyser/session/cache/SkullCache.java
#	core/src/main/java/org/geysermc/geyser/session/cache/WorldBorder.java
#	core/src/main/java/org/geysermc/geyser/translator/inventory/MerchantInventoryTranslator.java
#	core/src/main/java/org/geysermc/geyser/translator/level/block/entity/VaultBlockEntityTranslator.java
#	core/src/main/java/org/geysermc/geyser/translator/protocol/bedrock/entity/player/input/BedrockMovePlayer.java
#	core/src/main/java/org/geysermc/geyser/translator/protocol/java/entity/JavaAddEntityTranslator.java
#	core/src/main/java/org/geysermc/geyser/translator/protocol/java/entity/JavaEntityEventTranslator.java
#	core/src/main/java/org/geysermc/geyser/translator/protocol/java/entity/player/JavaPlayerInfoUpdateTranslator.java
#	core/src/main/java/org/geysermc/geyser/translator/protocol/java/entity/player/JavaPlayerPositionTranslator.java
#	core/src/main/java/org/geysermc/geyser/translator/protocol/java/level/JavaGameEventTranslator.java
#	core/src/test/java/org/geysermc/geyser/scoreboard/network/util/GeyserMockContextScoreboard.java

onebeastchris committed:
Fix hitbox querying, fix offset handling for boats

onebeastchris committed:
Fix entity lerping

onebeastchris committed:
Merge remote-tracking branch 'upstream/master' into feature/custom-entities-api

# Conflicts:
#	core/src/main/java/org/geysermc/geyser/entity/VanillaEntities.java
#	core/src/main/java/org/geysermc/geyser/entity/type/DisplayBaseEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/LivingEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/MinecartEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/TextDisplayEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/living/monster/ShulkerEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/player/AvatarEntity.java
#	core/src/main/java/org/geysermc/geyser/util/DimensionUtils.java

onebeastchris committed:
Merge remote-tracking branch 'upstream/master' into feature/custom-entities-api

onebeastchris committed:
Merge remote-tracking branch 'upstream/master' into feature/custom-entities-api

onebeastchris committed:
Remove bedrockPosition variable again, simpler to only store one source of truth

onebeastchris committed:
Merge remote-tracking branch 'upstream/master' into feature/custom-entities-api

# Conflicts:
#	core/src/main/java/org/geysermc/geyser/entity/VanillaEntities.java
#	gradle/libs.versions.toml

onebeastchris committed:
Address issues

onebeastchris committed:
Merge remote-tracking branch 'upstream/master' into feature/custom-entities-api

# Conflicts:
#	gradle/libs.versions.toml

onebeastchris committed:
Bump mcpl

onebeastchris committed:
Merge branch 'master' into feature/custom-entities-api

onebeastchris committed:
Merge remote-tracking branch 'upstream/master' into feature/custom-entities-api

# Conflicts:
#	.github/workflows/build.yml
#	api/src/main/java/org/geysermc/geyser/api/connection/GeyserConnection.java
#	api/src/main/java/org/geysermc/geyser/api/entity/EntityData.java
#	build-logic/src/main/kotlin/extensions.kt
#	gradle/libs.versions.toml

onebeastchris committed:
make it build again

onebeastchris committed:
Merge remote-tracking branch 'upstream/master' into feature/custom-entities-api

# Conflicts:
#	core/src/main/java/org/geysermc/geyser/session/cache/WorldBorder.java
#	core/src/main/java/org/geysermc/geyser/translator/protocol/bedrock/BedrockInventoryTransactionTranslator.java

onebeastchris committed:
Partially address review

onebeastchris committed:
Merge remote-tracking branch 'upstream/master' into feature/custom-entities-api

# Conflicts:
#	core/src/main/java/org/geysermc/geyser/entity/type/Entity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/TextDisplayEntity.java

onebeastchris committed:
fix le test

onebeastchris committed:
Merge branch 'master' into feature/custom-entities-api

onebeastchris committed:
Fix: ResourcePackClientResponsePacket's may contain empty pack lists in the send_packs case

onebeastchris committed:
Fix debug log formatting calls

onebeastchris committed:
Merge remote-tracking branch 'upstream/master' into feature/custom-entities-api

# Conflicts:
#	core/src/main/java/org/geysermc/geyser/translator/protocol/java/entity/JavaAddEntityTranslator.java
#	core/src/main/java/org/geysermc/geyser/translator/protocol/java/entity/player/JavaPlayerInfoUpdateTranslator.java
#	gradle/libs.versions.toml

onebeastchris committed:
Minor fixes, reviews, etc

onebeastchris committed:
Support setting empty hitbox

onebeastchris committed:
Merge branch 'master' into feature/custom-entities-api

onebeastchris committed:
Merge remote-tracking branch 'upstream/master' into feature/custom-entities-api

onebeastchris committed:
Reviews and comments

onebeastchris committed:
Improve height / width handling

This will be later reworked to allow setting a height + width combo that accounts for the entity pose. For now, it's a general override

onebeastchris commented on api/src/main/java/org/geysermc/geyser/api/entity/custom/CustomJavaEntityType.java:
```diff
@@ -0,0 +1,89 @@
+         * @param height the height
+         * @return this builder
+         */
+        Builder height(@NonNegative float height);
```
These should, ideally, allow registering height/width for different poses.

onebeastchris commented on core/src/main/java/org/geysermc/geyser/entity/properties/type/PropertyType.java:
```diff
@@ -38,7 +38,8 @@ public interface PropertyType<Type, NetworkRepresentation extends EntityProperty
 
-    default void apply(GeyserEntityPropertyManager manager, Type value) {
+    default void apply(@Nullable GeyserEntityPropertyManager manager, Type value) {
+        if (manager == null) return;
```
Probably worth adding a debug log here

onebeastchris committed:
Allow custom height / width to be 0

onebeastchris committed:
Merge remote-tracking branch 'upstream/master' into feature/custom-entities-api

onebeastchris committed:
Merge remote-tracking branch 'upstream/master' into feature/custom-entities-api

onebeastchris committed:
Likely fix entity registration

onebeastchris committed:
Merge branch 'master' into feature/custom-entities-api

onebeastchris committed:
Merge remote-tracking branch 'upstream/master' into feature/custom-entities-api

# Conflicts:
#	core/src/main/java/org/geysermc/geyser/entity/spawn/EntitySpawnContext.java
#	core/src/main/java/org/geysermc/geyser/entity/type/BoatEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/Entity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/EvokerFangsEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/ItemEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/ItemFrameEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/LightningEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/LivingEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/MinecartEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/TextDisplayEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/ThrowableEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/ThrowableItemEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/living/ArmorStandEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/living/animal/OcelotEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/living/animal/SnifferEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/living/merchant/VillagerEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/living/monster/CreakingEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/living/monster/EnderDragonEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/living/monster/EnderDragonPartEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/living/monster/EndermanEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/player/AvatarEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/player/SessionPlayerEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/vehicle/BoatVehicleComponent.java
#	core/src/main/java/org/geysermc/geyser/entity/vehicle/VehicleComponent.java
#	core/src/main/java/org/geysermc/geyser/inventory/holder/BlockInventoryHolder.java
#	core/src/main/java/org/geysermc/geyser/level/physics/CollisionManager.java
#	core/src/main/java/org/geysermc/geyser/session/cache/BlockBreakHandler.java
#	core/src/main/java/org/geysermc/geyser/session/cache/BossBar.java
#	core/src/main/java/org/geysermc/geyser/session/cache/WorldBorder.java
#	core/src/main/java/org/geysermc/geyser/translator/inventory/chest/DoubleChestInventoryTranslator.java
#	core/src/main/java/org/geysermc/geyser/translator/level/event/PlaySoundEventTranslator.java
#	core/src/main/java/org/geysermc/geyser/translator/protocol/bedrock/BedrockInventoryTransactionTranslator.java
#	core/src/main/java/org/geysermc/geyser/translator/protocol/bedrock/entity/player/input/BedrockMovePlayer.java
#	core/src/main/java/org/geysermc/geyser/translator/protocol/bedrock/entity/player/input/BedrockPlayerAuthInputTranslator.java
#	core/src/main/java/org/geysermc/geyser/translator/protocol/java/entity/JavaEntityEventTranslator.java
#	core/src/main/java/org/geysermc/geyser/translator/protocol/java/entity/player/JavaPlayerLookAtTranslator.java
#	core/src/main/java/org/geysermc/geyser/translator/protocol/java/entity/player/JavaPlayerPositionTranslator.java
#	core/src/main/java/org/geysermc/geyser/translator/protocol/java/level/JavaGameEventTranslator.java
#	core/src/main/java/org/geysermc/geyser/translator/protocol/java/level/JavaLevelParticlesTranslator.java
#	core/src/main/java/org/geysermc/geyser/util/InventoryUtils.java

onebeastchris committed:
Merge remote-tracking branch 'upstream/feature/custom-entities-api' into feature/custom-entities-api

onebeastchris committed:
Merge remote-tracking branch 'upstream/master' into feature/custom-entities-api

# Conflicts:
#	core/src/main/java/org/geysermc/geyser/entity/VanillaEntities.java

onebeastchris committed:
Reduce diff, fix issue related to odd entity movement

onebeastchris committed:
Merge remote-tracking branch 'upstream/master' into feature/custom-entities-api

onebeastchris committed:
Work around pistons creating ghost blocks

onebeastchris committed:
Merge branch 'master' into feature/custom-entities-api

onebeastchris committed:
Merge remote-tracking branch 'upstream/master' into feature/custom-entities-api

# Conflicts:
#	core/src/main/java/org/geysermc/geyser/session/GeyserSession.java
#	core/src/main/java/org/geysermc/geyser/translator/level/block/entity/PistonBlockEntity.java

onebeastchris committed:
Feature: Expose basic seat offset / update entity passengers event

onebeastchris committed:
Fix tests

onebeastchris committed:
Merge remote-tracking branch 'upstream/master' into feature/custom-entities-api

# Conflicts:
#	core/src/main/java/org/geysermc/geyser/entity/EntityDefinition.java

onebeastchris committed:
Merge remote-tracking branch 'upstream/master' into feature/custom-entities-api

# Conflicts:
#	core/src/main/java/org/geysermc/geyser/session/GeyserSession.java

onebeastchris committed:
Temporarily revert "Fix: hide living entity name when it is a vehicle (#6290)"

This reverts commit 3e9112967516928c5975e008f87aae89b550d1e2.

eclipseisoffline committed:
Fix: sent bogus furnace recipes when none are sent by the server

Remove debug exception

Fix: client crash because no furnace recipes are sent

# Conflicts:
#	core/src/main/java/org/geysermc/geyser/translator/protocol/java/JavaRecipeBookAddTranslator.java

onebeastchris committed:
Merge remote-tracking branch 'upstream/master' into feature/custom-entities-api

# Conflicts:
#	api/src/main/java/org/geysermc/geyser/api/connection/GeyserConnection.java
#	api/src/main/java/org/geysermc/geyser/api/entity/EntityData.java
#	api/src/main/java/org/geysermc/geyser/api/entity/type/GeyserEntity.java
#	core/src/main/java/org/geysermc/geyser/item/hashing/DataComponentHashers.java
#	core/src/main/java/org/geysermc/geyser/item/hashing/MinecraftHasher.java
#	core/src/main/java/org/geysermc/geyser/item/parser/ItemStackParser.java
#	core/src/main/java/org/geysermc/geyser/registry/Registries.java
#	core/src/main/java/org/geysermc/geyser/session/cache/EntityCache.java
#	core/src/main/java/org/geysermc/geyser/translator/protocol/java/JavaRecipeBookAddTranslator.java
#	gradle/libs.versions.toml

onebeastchris committed:
Update to 26.1

onebeastchris committed:
Update to jspecify conform nullability annotations

onebeastchris committed:
Merge remote-tracking branch 'upstream/master' into feature/custom-entities-api

# Conflicts:
#	core/src/main/java/org/geysermc/geyser/entity/VanillaEntities.java
#	core/src/main/java/org/geysermc/geyser/entity/type/living/merchant/VillagerEntity.java
#	core/src/main/java/org/geysermc/geyser/entity/type/player/SessionPlayerEntity.java
#	gradle/libs.versions.toml

onebeastchris committed:
Fix test, use a fixed snapshot build

onebeastchris committed:
Merge remote-tracking branch 'upstream/master' into feature/custom-entities-api

onebeastchris committed:
Merge remote-tracking branch 'refs/remotes/upstream/master' into feature/custom-entities-api

# Conflicts:
#	core/src/main/java/org/geysermc/geyser/level/physics/CollisionManager.java

onebeastchris committed:
Fix build

onebeastchris committed:
Implement new geyser entity data override system

onebeastchris committed:
Address reviews

onebeastchris committed:
Support more entity data types

onebeastchris committed:
Revert "Temporarily revert "Fix: hide living entity name when it is a vehicle (#6290)""

This reverts commit 60f0bb528f31b9e8357a02d009e70c11d114b41c.

onebeastchris committed:
Revert initial changes for supporting non-vanilla entities

onebeastchris committed:
le :sparkles: javadocs :sparkles:

onebeastchris committed:
Merge branch 'master' into feature/custom-entities-api

onebeastchris committed:
Merge remote-tracking branch 'upstream/master' into feature/custom-entities-api

onebeastchris committed:
Misc improvements: Javadocs, some structural fixes

onebeastchris committed:
Improve entity data system by using EntityDataBehaviorRegistry, improve some data type javadocs

onebeastchris committed:
Merge branch 'feature/26.2' into feature/custom-entities-api

# Conflicts:
#	core/src/main/java/org/geysermc/geyser/entity/VanillaEntities.java
#	core/src/main/java/org/geysermc/geyser/registry/Registries.java
#	core/src/main/java/org/geysermc/geyser/session/cache/registry/JavaRegistries.java
#	core/src/main/java/org/geysermc/geyser/translator/protocol/java/entity/JavaAddEntityTranslator.java
#	core/src/main/java/org/geysermc/geyser/translator/protocol/java/entity/player/JavaPlayerPositionTranslator.java

onebeastchris committed:
Fix: IdentifierImpl hacks

onebeastchris committed:
Mark API as experimental

onebeastchris committed:
Add debugging logs, smaller fixes to armor stand entity

onebeastchris committed:
Document remaining entity data types

onebeastchris committed:
this is aktschually unmodifiable

onebeastchris committed:
mhm this was supposed to be pushed

onebeastchris committed:
Update hitbox handling to resolve to-do

onebeastchris committed:
Differentiate between override and value in the API, improve offset handling in API

onebeastchris commented on api/src/main/java/org/geysermc/geyser/api/entity/data/types/Hitbox.java:
```diff
@@ -0,0 +1,127 @@
+import org.jetbrains.annotations.ApiStatus;
+
+/**
+ * Represents an entity hitbox, with min/max representing absolute coordinates
```
```suggestion
 * Represents an entity hitbox, with min/max representing absolute coordinates.
```

onebeastchris commented on api/src/main/java/org/geysermc/geyser/api/entity/data/types/Hitbox.java:
```diff
@@ -0,0 +1,127 @@
+public interface Hitbox {
+
+    /**
+     * The min "corner" of the hitbox as a position in the world
```
```suggestion
     * The min "corner" of the hitbox as a position in the world.
```

onebeastchris commented on api/src/main/java/org/geysermc/geyser/api/entity/data/types/Hitbox.java:
```diff
@@ -0,0 +1,127 @@
+    Vector3f min();
+
+    /**
+     * The max "corner" of the hitbox as a position in the world
```
```suggestion
     * The max "corner" of the hitbox as a position in the world.
```

onebeastchris commented on api/src/main/java/org/geysermc/geyser/api/entity/data/types/Hitbox.java:
```diff
@@ -0,0 +1,127 @@
+    Vector3f pivot();
+
+    /**
+     * Creates a new builder for a hitbox
```
```suggestion
     * Creates a new builder for a hitbox.
```

onebeastchris commented on api/src/main/java/org/geysermc/geyser/api/entity/data/types/Hitbox.java:
```diff
@@ -0,0 +1,127 @@
+    }
+
+    /**
+     * The builder for the hitbox
```
```suggestion
     * The builder for an entity hitbox.
```

onebeastchris commented on api/src/main/java/org/geysermc/geyser/api/entity/data/types/Hitbox.java:
```diff
@@ -0,0 +1,127 @@
+        @This Builder min(Vector3f min);
+
+        /**
+         * Sets the max corner of the hitbox
```
```suggestion
         * Sets the max corner of the hitbox.
```

onebeastchris commented on api/src/main/java/org/geysermc/geyser/api/entity/data/types/Hitbox.java:
```diff
@@ -0,0 +1,127 @@
+        @This Builder max(Vector3f max);
+
+        /**
+         * Sets the pivot of the hitbox
```
```suggestion
         * Sets the pivot of the hitbox.
```

onebeastchris commented on api/src/main/java/org/geysermc/geyser/api/entity/data/GeyserEntityDataType.java:
```diff
@@ -0,0 +1,115 @@
+ * Represents a type of entity data that can be sent for an entity.
+ * <p>
+ * Entity data types define the values stored for a particular piece of metadata,
+ * such as a {@code Byte}, {@code Integer}, {@code Float}; and the name associated with the types.
```
```suggestion
 * such as a {@code Byte}, {@code Integer}, {@code Float}; and the identifier associated with the types.
```

onebeastchris commented on api/src/main/java/org/geysermc/geyser/api/entity/data/GeyserEntityDataTypes.java:
```diff
@@ -0,0 +1,169 @@
+        GeyserEntityDataType.create(Identifier.of("seat_lock_rider_rotation_degrees"), Float.class);
+
+    /**
+     * Whether an entity riding another vehicle is sitting on a seat which has rotation.
```
```suggestion
     * Whether an entity riding another vehicle entity is sitting on a seat which has a rotation.
```

onebeastchris commented on api/src/main/java/org/geysermc/geyser/api/entity/data/GeyserEntityDataTypes.java:
```diff
@@ -0,0 +1,169 @@
+     */
+    @ApiStatus.Experimental
+    public static final GeyserEntityDataType<Boolean> SEAT_HAS_ROTATION =
+        GeyserEntityDataType.create(Identifier.of("seat_has_rotation"), Boolean.class);
```
```suggestion
        GeyserEntityDataType.create(Identifier.of("geysermc", "seat_has_rotation"), Boolean.class);
```

onebeastchris commented on api/src/main/java/org/geysermc/geyser/api/entity/definition/JavaEntityType.java:
```diff
@@ -0,0 +1,98 @@
+import org.jspecify.annotations.Nullable;
+
+/**
+ * Represents a Java edition entity type
```
```suggestion
 * Represents a Java edition entity type.
```

onebeastchris commented on api/src/main/java/org/geysermc/geyser/api/entity/type/GeyserEntity.java:
```diff
@@ -26,31 +26,140 @@
+
     /**
-     * @return the entity ID that the server has assigned to this entity.
+     * @return the entity ID that the server has assigned to this entity, or 0 if none is present
```
```suggestion
     * @return the entity ID that the server has assigned to this entity, or -1 if none is present
```

onebeastchris commented on api/src/main/java/org/geysermc/geyser/api/entity/type/GeyserEntity.java:
```diff
@@ -26,31 +26,140 @@
+    long geyserId();
+
+    /**
+     * @return the entity UUID that the server has assigned to this entity,
```
```suggestion
     * The entity's UUID, only null if the server has not spawned this entity.
     *
     * @return the entity UUID that the server has assigned to this entity,
```

onebeastchris commented on api/src/main/java/org/geysermc/geyser/api/entity/type/GeyserEntity.java:
```diff
@@ -26,31 +26,140 @@
+    UUID uuid();
+
+    /**
+     * The Bedrock entity definition for this entity, which could also be a {@link CustomEntityDefinition}
```
```suggestion
     * The Bedrock entity definition for this entity, which can also be a {@link CustomEntityDefinition}.
```

onebeastchris commented on api/src/main/java/org/geysermc/geyser/api/entity/EntityData.java:
```diff
@@ -29,24 +29,57 @@
- * This class holds all the methods that relate to entities.
- * Can be accessed through {@link GeyserConnection#entities()}.
+ * Provides entity lookup and input-lock utilities for a specific connection.
+ * Accessible via {@link GeyserConnection#entities()}.
```
```suggestion
 * Accessed via {@link GeyserConnection#entities()}.
```

onebeastchris commented on api/src/main/java/org/geysermc/geyser/api/event/lifecycle/GeyserDefineEntitiesEvent.java:
```diff
@@ -0,0 +1,81 @@
+     * Returns an immutable view of all currently registered Bedrock entity definitions,
+     * including both vanilla and any custom definitions registered so far in this event.
+     *
+     * @return immutable collection of all registered entity definitions
```
```suggestion
     * @return an immutable collection of all registered entity definitions
```

onebeastchris commented on api/src/main/java/org/geysermc/geyser/api/event/lifecycle/GeyserDefineEntitiesEvent.java:
```diff
@@ -0,0 +1,81 @@
+    /**
+     * Returns an immutable view of all custom entity definitions registered so far.
+     *
+     * @return immutable collection of registered custom entity definitions
```
```suggestion
     * @return an immutable collection of registered custom entity definitions
```

onebeastchris commented on api/src/main/java/org/geysermc/geyser/api/event/lifecycle/GeyserDefineEntitiesEvent.java:
```diff
@@ -0,0 +1,81 @@
+     *
+     * @param definition the custom entity definition to register; must not be null
+     * @throws IllegalArgumentException if {@code definition} was not created via {@link CustomEntityDefinition#of}
+     * @throws IllegalStateException if the identifier has already been registered
```
```suggestion
     * @throws IllegalStateException if a custom entity definition with this identifier has already been registered
```

onebeastchris committed:
Include number of custom entities in the API

onebeastchris committed:
Merge remote-tracking branch 'upstream/feature/26.2' into feature/custom-entities-api

# Conflicts:
#	core/src/main/java/org/geysermc/geyser/entity/type/Entity.java
#	core/src/main/java/org/geysermc/geyser/registry/Registries.java
#	core/src/main/java/org/geysermc/geyser/registry/loader/ProviderRegistryLoader.java
#	core/src/main/java/org/geysermc/geyser/session/GeyserSession.java

onebeastchris committed:
Apply suggestions from code review

Co-authored-by: chris <github@onechris.mozmail.com>

onebeastchris committed:
Merge fixes

onebeastchris committed:
Merge remote-tracking branch 'upstream/feature/custom-entities-api' into feature/custom-entities-api

onebeastchris committed:
Add read/write lock for entity cache to allow safe off-thread entity lookups

onebeastchris committed:
dirtyMetadata -> metadata

The class is no longer just a wrapper for metadata, but also handles API overrides / tracks entity data exposed in the API

onebeastchris committed:
Apply suggestion
